# Taller 00 - Analítica de Datos
- **Autor:** David Santiago Barreto Mora.
- **Fecha de entrega:** 10/08/2023 
- **Profesor:** Ing. John Corredo, PhD.
---

## Fuente de datos
Para el desarrollo del taller, se usan los archivos de extension .csv dados por el profesor. Los archivos detallan las ventas hechas en Colombia en 2019, de lo que se presume es una tienda de tecnología. En total son 12 archivos, cada uno correspondiendo a las ventas de un mes, por lo que el dataset en total constituye las ventas anuales del 2019 de la empresa en cuestión.

Para poder comenzar realizar el trabajo analítico, primero importamos las librerias a ser usadas, y por supuesto, los archivos .csv.



### Carga de librerias


In [2]:
# Se importan las librerias que serán usadas en el análisis.
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib as plt


### Carga de los datos (archivos .csv) 
A continuación se cargan los 12 archivos CSV y se crea un dataframe por cada uno. Esto con la intención de unir todos los dataframes en uno solo, para posteriormente facilitar su manipulación y análisis.

In [3]:
# Se importan los 12 archivos .csv, y por cada uno se crea un dataframe (libreria pandas)
dfJanuary = pd.read_csv("Ventas/Ventas_January_2019.csv")
dfFebruary = pd.read_csv("Ventas/Ventas_February_2019.csv")
dfMarch = pd.read_csv("Ventas/Ventas_March_2019.csv")
dfApril = pd.read_csv("Ventas/Ventas_April_2019.csv")
dfMay = pd.read_csv("Ventas/Ventas_May_2019.csv")
dfJune = pd.read_csv("Ventas/Ventas_June_2019.csv")
dfJuly = pd.read_csv("Ventas/Ventas_July_2019.csv")
dfAugust = pd.read_csv("Ventas/Ventas_August_2019.csv")
dfSeptember = pd.read_csv("Ventas/Ventas_September_2019.csv")
dfOctober = pd.read_csv("Ventas/Ventas_October_2019.csv")
dfNovember = pd.read_csv("Ventas/Ventas_November_2019.csv")
dfDecember = pd.read_csv("Ventas/Ventas_December_2019.csv")


A continuación imprimimos las 5 primeras fila de cada dataframe. Principalmente, para ver que las columnas de todos los dataframes sean las mismas. Esta verificicación es hecha para validar que se puedan unir los dataframes en uno solo.

In [4]:
# Se muestra la primera columna del dataframe de enero.
dfJanuary.head(5)


ID Orden                 Producto   Cantidad Ordenada Precio Unitario   \
0   141234   Audifonos Apple Airpods                  1           499900   
1   141235  Parlante Google HomeMini                  1           159900   
2   141236  Cable de Carga Lightning                  1            64900   
3   141237  Parlante Google HomeMini                  1           159900   
4   141238       Samsung Galaxy A302                  1           649900   

            Fecha              Dirección de Compra  
0  01/25/19 13:51      177 calle Jimenez, Cali, VC  
1  01/12/19 23:58        732 calle 9na, Cúcuta, NS  
2  01/15/19 21:42  887 calle Kennedy, Medellin, AN  
3  01/27/19 11:39   641 calle 5ta, Bucaramanga, SA  
4  01/14/19 17:13      379 calle 2da, Medellin, AN

In [5]:
# Se muestra la primera columna del dataframe de febrero.
dfFebruary.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   153258      Multiolla Oster 0150 Negra PE                  1   
1   153259              AA Batteries (4-pack)                  2   
2   153260  Horno Microondas 0.8P Samsung AGE                  1   
3   153261           Cable de Carga Lightning                  1   
4   153262           Parlante Google HomeMini                  1   

  Precio Unitario            Fecha              Dirección de Compra  
0           299900  02/04/19 07:28      724 calle Adams, Cúcuta, NS  
1            13800  02/22/19 20:19      61 calle Pinar, Armenia, QU  
2           309900  02/12/19 12:34   40 calle Jimenez, Medellin, AN  
3            64900  02/09/19 18:07  670 calle Jefferson, Bogotá, CU  
4           159900  02/05/19 18:34    223 calle Boyacá, Armenia, QU

In [6]:
# Se muestra la primera columna del dataframe de marzo.
dfMarch.head(5)


ID Orden                  Producto   Cantidad Ordenada Precio Unitario   \
0   163171  LG Secadora Carga Frontal                  1          1646900   
1   163172      AA Batteries (4-pack)                  1            13800   
2   163173  Audifonos Bose SoundSport                  1           399900   
3   163174   Cable de Carga USB-C P/O                  2            46900   
4   163175   Cable de Carga Lightning                  1            64900   

            Fecha                      Dirección de Compra  
0  03/31/19 11:58            591 calle Adams, Medellin, AN  
1  03/23/19 13:24             106 calle Adams, Armenia, QU  
2  03/08/19 13:46            545 calle Rivera, Armenia, QU  
3  03/30/19 18:04  370 calle Autopista Norte, Medellin, AN  
4  03/06/19 21:08              347 calle Sucre, Ibagué, TO

In [7]:
# Se muestra la primera columna del dataframe de abril
dfApril.head(5)


ID Orden                 Producto   Cantidad Ordenada Precio Unitario   \
0   171346  Cable de Carga USB-C P/O                  1            46900   
1   171347  Parlante Google HomeMini                  1           159900   
2   171348    AAA Batteries (4-pack)                  1            12990   
3   171349  Cable de Carga USB-C P/O                  1            46900   
4   171350  Cable de Carga USB-C P/O                  1            46900   

            Fecha            Dirección de Compra  
0  04/04/19 19:25  631 calle Boyacá, Armenia, QU  
1  04/19/19 07:19    546 calle 5ta, Medellin, AN  
2  04/22/19 22:41   593 calle 8va, Manizalez, CA  
3  04/09/19 04:59    571 calle 4ta, Medellin, AN  
4  04/18/19 09:36   883 calle Madison, Neiva, HU

In [8]:
# Se muestra la primera columna del dataframe de mayo.
dfMay.head(5)


ID Orden                  Producto   Cantidad Ordenada Precio Unitario   \
0   182565  LG Lavadora 19Kg Inv WT19                  1          1646900   
1   182566        Xiaomi REDMI Note 8                  1           719000   
2   182567      AA Batteries (4-pack)                  1            13800   
3   182568   Cable de Carga Lightning                  1            64900   
4   182569   Parlante Google HomeMini                  1           159900   

            Fecha            Dirección de Compra  
0  05/05/19 10:02    722 calle Adams, Bogotá, CU  
1  05/04/19 22:44  986 calle Johnson, Bogotá, CU  
2  05/02/19 17:24     422 calle Rio, Armenia, QU  
3  05/25/19 22:49      630 calle 7ma, Bogotá, CU  
4  05/20/19 17:24   741 calle 7ma, Manizalez, CA

In [9]:
# Se muestra la primera columna del dataframe de junio.
dfJune.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   195098           Cable de Carga Lightning                  1   
1   195099           Cable de Carga Lightning                  1   
2   195100  Freidora Air Fryer Kalley K-MAF25                  1   
3   195101             AAA Batteries (4-pack)                  1   
4   195102              AA Batteries (4-pack)                  1   

  Precio Unitario            Fecha                  Dirección de Compra  
0            64900  06/17/19 20:12  736 calle Transversal, Medellin, AN  
1            64900  06/25/19 00:27          562 calle Elm, Medellin, AN  
2           198849  06/23/19 20:44        548 calle Sucre, Medellin, AN  
3            12990  06/12/19 20:30          322 calle Sur, Medellin, AN  
4            13800  06/02/19 10:00        928 calle Avenida, Ibagué, TO

In [10]:
# Se muestra la primera columna del dataframe de julio.
dfJuly.head(5)


ID Orden                       Producto   Cantidad Ordenada  \
0   211128          AAA Batteries (4-pack)                  1   
1   211129  Asus VivoBook X420FA-BV Laptop                  1   
2   211130   Multiolla Oster 0150 Negra PE                  1   
3   211131           AA Batteries (4-pack)                  1   
4   211132           AA Batteries (4-pack)                  1   

  Precio Unitario            Fecha                 Dirección de Compra  
0            12990  07/09/19 21:48       327 calle Carrera, Cúcuta, NS  
1          1950900  07/14/19 15:56        597 calle Rio, Manizalez, CA  
2           299900  07/28/19 21:25          658 calle 14va, Ibagué, TO  
3            13800  07/24/19 12:47  100 calle Transversal, Armenia, QU  
4            13800  07/20/19 18:30         494 calle Elm, Medellin, AN

In [11]:
# Se muestra la primera columna del dataframe de agosto.
dfAugust.head(5)


ID Orden                 Producto   Cantidad Ordenada Precio Unitario   \
0   224782    AAA Batteries (4-pack)                  2            12990   
1   224783    AAA Batteries (4-pack)                  1            12990   
2   224784     AA Batteries (4-pack)                  1            13800   
3   224785  Cable de Carga Lightning                  1            64900   
4   224786    AAA Batteries (4-pack)                  1            12990   

            Fecha             Dirección de Compra  
0  08/04/19 08:24    91 calle Boyacá, Armenia, QU  
1  08/31/19 07:55  157 calle Willow, Medellin, AN  
2  08/30/19 18:25   317 calle Santa, Medellin, AN  
3  08/13/19 13:48      744 calle Centro, Cali, VC  
4  08/08/19 10:52   486 calle Cherry, Armenia, QU

In [12]:
# Se muestra la primera columna del dataframe de septiembre.
dfSeptember.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   242703                Samsung Galaxy A302                  1   
1   242704           Cable de Carga Lightning                  1   
2   242705          Audifonos Bose SoundSport                  1   
3   242706  Horno Microondas 0.8P Samsung AGE                  1   
4   242707  Freidora Saludable B&D 3.5 Litros                  1   

  Precio Unitario            Fecha                  Dirección de Compra  
0           649900  09/19/19 12:57        508 calle 11ra, Manizalez, CA  
1            64900  09/10/19 20:36          401 calle 9na, Medellin, AN  
2           399900  09/24/19 18:44  320 calle Portal Sur, Manizalez, CA  
3           309900  09/23/19 12:19      760 calle 11ra, Bucaramanga, SA  
4           229900  09/03/19 00:09        5 calle Diagonal, Armenia, QU

In [13]:
# Se muestra la primera columna del dataframe de octubre.
dfOctober.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   246400  Freidora Saludable B&D 3.5 Litros                  1   
1   246401           Parlante Google HomeMini                  1   
2   246402              AA Batteries (4-pack)                  1   
3   246403           Cable de Carga USB-C P/O                  1   
4   246404           Cable de Carga USB-C P/O                  1   

  Precio Unitario            Fecha                   Dirección de Compra  
0           229900  10/05/19 17:01         893 calle 14va, Manizalez, CA  
1           159900  10/09/19 13:20         502 calle Chapinero, Cali, VC  
2            13800  10/19/19 13:04          202 calle Cedar, Armenia, QU  
3            46900  10/21/19 03:12  954 calle Americas, Barranquilla, AT  
4            46900  10/10/19 21:15           79 calle Sunset, Ibagué, TO

In [14]:
# Se muestra la primera columna del dataframe de noviembre.
dfNovember.head(5)


ID Orden                          Producto   Cantidad Ordenada  \
0   260944  Freidora Saludable B&D 3.5 Litros                  1   
1   260945            Audifonos Apple Airpods                  1   
2   260946           Cable de Carga USB-C P/O                  2   
3   260947          Audifonos Bose SoundSport                  1   
4   260948            Audifonos Apple Airpods                  1   

  Precio Unitario            Fecha                     Dirección de Compra  
0           229900  11/12/19 14:57           79 calle Avenida, Armenia, QU  
1           499900  11/04/19 17:52           849 calle Atalaya, Bogotá, CU  
2            46900  11/20/19 11:37                99 calle 9na, Ibagué, TO  
3           399900  11/26/19 17:58  126 calle Portal Sur, Barranquilla, AT  
4           499900  11/27/19 18:16           983 calle Museo, Medellin, AN

In [15]:
# Se muestra la primera columna del dataframe de diciembre.
dfDecember.head(5)


ID Orden                      Producto   Cantidad Ordenada Precio Unitario   \
0   276339      Audifonos Bose SoundSport                  1           399900   
1   276340  Multiolla Oster 0150 Negra PE                  1           299900   
2   276341            Samsung Galaxy A302                  1           649900   
3   276342       Cable de Carga Lightning                  1            64900   
4   276343      Audifonos Bose SoundSport                  1           399900   

            Fecha                Dirección de Compra  
0  12/11/19 19:29    345 calle Pte Aranda, Neiva, HU  
1  12/13/19 16:56  924 calle Americas, Manizalez, CA  
2  12/12/19 06:05     616 calle Sunset, Medellin, AN  
3  12/23/19 21:22     189 calle Willow, Medellin, AN  
4  12/20/19 15:36      931 calle Madison, Bogotá, CU

Como se puede ver, todos los dataframes (o archivos .csv) cuentan con 6 columnas:
1. ID Orden
2. Producto
3. Cantidad Ordenada
4. Precio Unitario
5. Fecha
6. Dirección de Compra

### 2.C Unión de los dataframes 
Para poder consolidar todos los datos en un solo archivo, es necesario entender como funciona la función merge (ya que es requerido unir los dataframes con esta función). 

Si nos guiamos por la documentación oficial, la función es casi que análoga a un JOIN de SQL. Esto es supremamente importante de entender, porque la tabla resultante de un JOIN puede tener nuevas filas o columnas que no estaban previamente en ninguna de las 2 tablas usadas. Lo que nosotros buscamos en este análisis es anexar los datos de una tabla "debajo" de otra, con el fin de poder tener toda la información de los 12 meses sin alteración alguna y de manera continua; lo cual es problemático teniendo en cuenta las nuevas filas que puede generar un join. 

A continuación una demostración del problema que queremos evitar. Supongamos que vamos a usar 'merge()' con las tablas de Enero y Febrero. Primero mostramos la cantidad de filas y columnas de cada dataframe.

In [16]:
print("Dimensiones del Dataframe de Enero:", dfJanuary.shape)
print("Dimensiones del Dataframe de Enero:", dfFebruary.shape)

Dimensiones del Dataframe de Enero: (12554, 6)
Dimensiones del Dataframe de Enero: (10355, 6)


Podemos ver que las tablas tienen 12554 y 10355 filas, respectivamente. Por ahora solo nos interesa el numero de filas.

Si ahora ejecutamos la función merge con los dos dataframes:

In [17]:
testMergedDf = pd.merge(dfJanuary, dfFebruary, how="outer")
print("Dimensiones del Dataframe de Enero:", testMergedDf.shape)
print(12554 + 10355)

Dimensiones del Dataframe de Enero: (24474, 6)
22909


Podemos ver que el dataframe combinado tiene en total 24474 filas, cuando la suma de las filas de las tablas originales da 22909. Más de 1000 registros nuevos fueron creados. También resalto el parámetro "how", el cual especifica como unir las tablas (inner join, outer join, left outer join, etc). El ejemplo se muestra con "outer" porque esto es lo más parecido a unir totalmente las 2 tablas. 

Debido a lo anterior, usaremos la función ***concat ( )***, la cual nos permite adjuntar las columnas de los dataframes en la forma deseada. También porque es una versión actualizada de la función ***append( )***, la cual pandas nos advierte ya está obsoleta y será removida a futuro. Primero imprimimos la cantidad de filas de todos los dataframes para verificar que se esten unificando apropiadamente.



In [18]:
print("Enero: ", dfJanuary.shape)
print("Febrero: ", dfFebruary.shape)
print("Marzo: ",dfMarch.shape)
print("Abril: ", dfApril.shape)
print("Mayo: ",dfMay.shape)
print("Junio: ",dfJune.shape)
print("Julio: ",dfJuly.shape)
print("Agosto: ",dfAugust.shape)
print("Septiembre: ", dfSeptember.shape)
print("Octubre: ",dfOctober.shape)
print("Noviembre: ", dfNovember.shape)
print("Diciembre: ", dfDecember.shape)
print("Total: ", (12554 + 10355 + 8561 + 11754 + 13074 + 16717 + 14271 +18702 + 3842 + 15184 + 16057 + 24132))

Enero:  (12554, 6)
Febrero:  (10355, 6)
Marzo:  (8561, 6)
Abril:  (11754, 6)
Mayo:  (13074, 6)
Junio:  (16717, 6)
Julio:  (14271, 6)
Agosto:  (18702, 6)
Septiembre:  (3842, 6)
Octubre:  (15184, 6)
Noviembre:  (16057, 6)
Diciembre:  (24132, 6)
Total:  165203


Y ahora adjuntamos todos los dataframes simultaneamente.

In [19]:
ventasAnuales = pd.concat([dfJanuary, dfFebruary, dfMarch, dfApril, dfMay, dfJune, dfJuly, dfAugust, dfSeptember, dfOctober, dfNovember, dfDecember])
print(ventasAnuales.shape)

(165203, 6)


Como podemos ver, la cantida de filas del nuevo dataframe corresponde con la suma hecha previamente.


### 2.D Guardado del consolidado en un solo fichero.
El guardado lo podemos realizar con la función ***to_csv()*** también dada por pandas.

In [20]:
ventasAnuales.to_csv("ventas_anules.csv")

## 2.E Análisis con el dataframe consolidado

### 2.E.1 Imprima los primeros 10 registros.
Hacemos uso de la función ***df.head()***, especificando 10 registros en el argumento de la función.

In [21]:
ventasAnuales.head(10)

ID Orden                      Producto   Cantidad Ordenada Precio Unitario   \
0   141234        Audifonos Apple Airpods                  1           499900   
1   141235       Parlante Google HomeMini                  1           159900   
2   141236       Cable de Carga Lightning                  1            64900   
3   141237       Parlante Google HomeMini                  1           159900   
4   141238            Samsung Galaxy A302                  1           649900   
5   141239       Parlante Google HomeMini                  1           159900   
6   141240         AAA Batteries (4-pack)                  3            12990   
7   141241       Parlante Google HomeMini                  2           159900   
8   141242  Multiolla Oster 0150 Negra PE                  1           299900   
9   141243       Cable de Carga USB-C P/O                  1            46900   

            Fecha               Dirección de Compra  
0  01/25/19 13:51       177 calle Jimenez, Cali, VC  
1  01/12/19 23:58         732 calle 9na, Cúcuta, NS  
2  01/15/19 21:42   887 calle Kennedy, Medellin, AN  
3  01/27/19 11:39    641 calle 5ta, Bucaramanga, SA  
4  01/14/19 17:13       379 calle 2da, Medellin, AN  
5  01/24/19 10:32         92 calle Santa, Neiva, HU  
6  01/30/19 14:44         618 calle Pinar, Cali, VC  
7  01/12/19 12:41   128 calle Lincoln, Medellin, AN  
8  01/07/19 18:32  182 calle Portal Sur, Bogotá, CU  
9  01/07/19 16:45          266 calle 14va, Cali, VC

### 2.E.2. Imprima los últimos 5 registros.
De forma similar, hacemos uso de la función ***df.tail()***.

In [22]:
ventasAnuales.tail(5)

ID Orden                  Producto   Cantidad Ordenada Precio Unitario   \
24127   299477  Audifonos Bose SoundSport                  1           399900   
24128   299478   Cable de Carga Lightning                  1            64900   
24129   299479   Cable de Carga USB-C P/O                  1            46900   
24130   299480     AAA Batteries (4-pack)                  1            12990   
24131   299481   Cable de Carga USB-C P/O                  1            46900   

                Fecha                     Dirección de Compra  
24127  12/01/19 10:38  421 calle Autopista Norte, Armenia, QU  
24128  12/27/19 05:23          679 calle 4ta, Bucaramanga, SA  
24129  12/15/19 12:41           707 calle Norte, Medellin, AN  
24130  12/05/19 16:21         735 calle Carrera, Medellin, AN  
24131  12/19/19 01:09      175 calle Portal Sur, Medellin, AN

### 2.E.4 Cuales y cuantos tipos de datos se encuentran en el dataset.
En esta ocasión utilazamos la función ***df.info()***.

In [23]:
ventasAnuales.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 165203 entries, 0 to 24131
Data columns (total 6 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   ID Orden             164707 non-null  object
 1   Producto             164707 non-null  object
 2    Cantidad Ordenada   164707 non-null  object
 3   Precio Unitario      164707 non-null  object
 4   Fecha                164707 non-null  object
 5   Dirección de Compra  164707 non-null  object
dtypes: object(6)
memory usage: 8.8+ MB


### 2.E.3 Presente la información del dataset y haga sus observaciones.
Habiendo presentado los tipos de datos dentro del dataframe, ahora podemos hacer algunas observaciones sobre el dataset.

Basandonos en los puntos 2.E.1 y 2.E.2 podemos ver que el dataset contiene información sobre las ventas hechas de un negocio aparentemente colombiano. Dentro del dataset encontramos la información de cada venta hecha; cada orden conteniendo producto vendido, cantidad de producto, precio unitario y dirección.

Algo también interesante es la cantidad de datos nulos en el dataset:

In [24]:
print(ventasAnuales.shape)
print(165203 - 164707)

(165203, 6)
496


### 2.E.5 Justifique y elimine datos nulos.
De las 165,203 fila del dataframe, la función info() nos dice que 164,707 tienen la información completa. Esto significa que solo hay 496 filas nulas. Estas 496 filas solo representtan un 0.3% del dataset, por lo que se toma la decisión de borrar las columnas.

La decisión fue hecha basada principalmente sobre la insignificancia estadística que tienen las filas borradas. Las filas nulas constituyen una fracción insignificante de todo el conjunto de datos y es poco probable que su eliminación introduzca sesgos o altere significativamente las propiedades generales de los datos. En consecuencia, no se espera que la eliminación de estas filas comprometa la integridad del análisis.

Adicionalmente, también podemos ver que todas las demas filas del dataframe tienen sus datos completos; no hay ni una sola fila con una celda nula. Podríamos asumir que las filas nulas son resultado de errores triviales a la hora de introducir informacion al dataset o en su translado. Eliminarlas solo fortalece la calidad de los datos presentes.

In [25]:
ventasAnuales_2 = ventasAnuales.dropna()
print(ventasAnuales_2.shape)

(164707, 6)


### 2.E.6 Imprima el nombre de las columnas del objeto dataframe.
Todos los dataframes cuentan con el atributo ***.columns*** el cual lista el nombre de todas sus columnas.

In [26]:
ventasAnuales.columns

Index(['ID Orden', 'Producto ', ' Cantidad Ordenada', 'Precio Unitario ',
       'Fecha', 'Dirección de Compra'],
      dtype='object')

En caso de que haya duda, esto nos verifica que el dataframe solo cuenta con 6 columnas:
1. ID Orden
2. Producto
3. Cantidad Ordenada
4. Precio Unitario
5. Fecha
6. Dirección de Compra

### 2.E.7 Haga almenos 2 cambios de nombre de columnas y justifique. 
Se decide renombrar aquellas columnas que tengan espacios en sus nombres. La decisión fue fue impulsada con el objetivo de mejorar la claridad y legibilidad del conjunto de datos. Al adoptar nombres de columna descriptivos y coherentes, mejoramos la comprensión de los datos y facilitamos un análisis más eficaz de los mismos. Los nuevos nombres de las columnas n solo ofrecen una representación concisa de la información contenida en cada una de ellas, sino que también nos permiten evitar posibles problemas que surgan de sus nombres.


In [27]:
nuevosNombresColumnas = {
    'ID Orden': 'ID_Orden',
    'Producto ': 'Producto',
    ' Cantidad Ordenada': 'CantidadOrdenada',
    'Precio Unitario ': 'PrecioUnitario',
    'Fecha': 'Fecha',
    'Dirección de Compra': 'DireccionCompra'
}

ventasAnuales_2.rename(columns=nuevosNombresColumnas, inplace=True)
ventasAnuales_2.head()

C:\Users\David Barreto\AppData\Local\Temp\ipykernel_22600\2395415418.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ventasAnuales_2.rename(columns=nuevosNombresColumnas, inplace=True)


ID_Orden                  Producto CantidadOrdenada PrecioUnitario  \
0   141234   Audifonos Apple Airpods                1         499900   
1   141235  Parlante Google HomeMini                1         159900   
2   141236  Cable de Carga Lightning                1          64900   
3   141237  Parlante Google HomeMini                1         159900   
4   141238       Samsung Galaxy A302                1         649900   

            Fecha                  DireccionCompra  
0  01/25/19 13:51      177 calle Jimenez, Cali, VC  
1  01/12/19 23:58        732 calle 9na, Cúcuta, NS  
2  01/15/19 21:42  887 calle Kennedy, Medellin, AN  
3  01/27/19 11:39   641 calle 5ta, Bucaramanga, SA  
4  01/14/19 17:13      379 calle 2da, Medellin, AN

### 2.E.8 Presente un informe estadístico con observaciones
A continuación se presentan algunos puntos estadísticos relevantes del informe.


In [ ]:
ventasAnuales_2['Fecha'] = pd.to_datetime(ventasAnuales_2['Fecha'])

# Create a bar plot for top-selling products
top_products = ventasAnuales_2['Producto'].value_counts().head(10)
plt.figure(figsize=(10, 6))
sns.barplot(x=top_products.values, y=top_products.index, palette='viridis')
plt.title('Top 10 Selling Products')
plt.xlabel('Number of Orders')
plt.ylabel('Product')
plt.tight_layout()
plt.show()

# Create a time series plot for product sales over time
plt.figure(figsize=(12, 6))
product_sales_over_time = ventasAnuales_2.groupby('Fecha')['CantidadOrdenada'].sum()
product_sales_over_time.plot(marker='o')
plt.title('Product Sales Over Time')
plt.xlabel('Fecha')
plt.ylabel('Total Quantity Ordered')
plt.tight_layout()
plt.show()

### 2.E.9 Haga un estudio de datos atípicos y aplique un metodo para eliminarlos
El proceso de eliminación de valores atípicos se empleó para aumentar la solidez y fiabilidad del conjunto de datos para el análisis. Los valores atípicos, o puntos de datos extremos que se desvían significativamente de la distribución típica, tienen el potencial de distorsionar las medidas estadísticas y llevar a conclusiones inexactas. Para abordar este problema, se estableció un umbral de 3 veces la desviaciones estándar de la media.

Al fijar este umbral, pretendíamos identificar y excluir los puntos de datos que presentaban una desviación sustancial del valor medio. El método elegido preserva la integridad de la mayoría de los datos al tiempo que mitiga la influencia de los valores extremos que, de otro modo, podrían introducir sesgos o distorsiones en los análisis posteriores.

Este método de eliminación de valores atípicos se ajusta a las prácticas estadísticas establecidas y garantiza que nuestro análisis se centre en las tendencias y pautas fundamentales presentes en los datos. Como resultado, se espera que los conocimientos derivados del conjunto de datos sean más precisos, informativos y representativos de los fenómenos subyacentes.

In [ ]:
ventasAnuales_2['CantidadOrdenada'] = pd.to_numeric(ventasAnuales_2['Quantity_Ordered'], errors='coerce')

# Calculate summary statistics
mean = ventasAnuales_2['CantidadOrdenada'].mean()
std_dev = ventasAnuales_2['CantidadOrdenada'].std()

# Set threshold for outlier removal (e.g., 3 standard deviations from the mean)
threshold = 3 * std_dev

# Identify and remove outliers
df_cleaned = ventasAnuales_2[np.abs(ventasAnuales_2['CantidadOrdenada'] - mean) <= threshold]

# Justification for Outlier Removal:
justification = (
    "Outliers were removed using a threshold of {} standard deviations from the mean. "
    "This method helps eliminate extreme data points that could potentially skew "
    "the analysis results while retaining the majority of the data."
).format(threshold)

# Display the cleaned DataFrame
print(df_cleaned.head())
print(justification)